meteor-bundle-main.js

```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
           // hasn't given us a signed certificate yet, the response-json
           // won't exist.
           if ((e.code === 'EACCES') || (e.code == 'ENOENT')) {
-            console.log("Skipping unreadable HTTPS key information file:", metadataFilename);
+            console.log("Skipping unreadable HTTPS key information file while examining:",
+                        metadataFilename);
             continue;
           } else {
-            throw e;
+            // Sometimes the server gives us a 0-byte response,
+            // presumably due to a connection getting reset?.
+            console.error("Got exception reading JSON from file:", metadataFilename, "Dazed and confused, but trying to continue.", e);
+            continue;
           }
         }
 
-        // If this certificate isn't valid yet, keep looping, hoping
-        // to find one that is valid.
-        var notBefore = metadata['notBefore'];
-        if (notBefore && (notBefore < nowUnixTimestamp)) {
+        // If the cert is expired, definitely don't use it.
+        if (now > metadata.notAfter) {
+          console.log("Skipping", keyFilename, "because", metadata.notAfter, "is in the past.");
           continue;
         }
 
-        result['ca'] = metadata['ca'];
-        result['key'] = fs.readFileSync(keyFilename, 'utf-8');
-        result['cert'] = metadata['cert'];
-        return result;
+        validCertificates.push(metadata);
+      }
+
+      // Sort by notBefore, ascending.
+      validCertificates.sort(function(cert1, cert2) {
+        return (cert1.notBefore - cert2.notBefore);
+      });
+
+      if (! validCertificates) {
+        // TODO: Be more robust here, though it's not clear how.
+        console.error("Aieee, we found zero valid certificates. Allowing Sandstorm to crash.");
+        return;
       }
+
+      console.log("Using this HTTPS key:", validCertificates[0].keyFilename,
+                  "valid starting", new Date(validCertificates[0].notBefore),
+                  "until", new Date(validCertificates[0].notAfter));
+
+      // Store HTTPS configuration data from the oldest certificate.
+      var result = {
+        ca: validCertificates[0].ca,
+        key: fs.readFileSync(validCertificates[0].keyFilename, 'utf-8'),
+        cert: validCertificates[0].cert
+      };
+
+      // Calculate re-key time.
+      //
+      // - If the cert we want to switch to is not valid yet, then we
+      //   re-key at its notBefore time.
+      //
+      // - If the cert we want to switch to *is* already valid, then
+      //   we re-key at the notAfter time of our current cert.
```

Comment:
It'd be nice if we could always rekey immediately, but I prefer my current algorithm because it's easier for me to explain to myself how it operates in the face of clock skew. Can think about this more, but would prefer to do so on a separate pull request if the current version is not wrong.